synapse/rest/client/transactions.py

```diff
@@ -81,7 +81,16 @@ def fetch_or_execute(self, txn_key, fn, *args, **kwargs):
             Deferred which resolves to a tuple of (response_code, response_dict).
         """
         try:
-            return self.transactions[txn_key][0].observe()
+            observable = self.transactions[txn_key][0]
+            if not observable.has_called() or observable.has_succeeded():
+                return observable.observe()
+            # if the request has already been called with a non-2xx status
+            # (a Twisted failure), remove it from the transaction map.
+            # This is done to ensure that we don't cache rate-limiting errors, etc.
+            res = observable.get_result()
+            if res.value.code >= 300:
```

Comment:
What should I do to check if there is a `code` attribute? Try to cast to `CodeMessageException`?